DBZ-993 Document schema update process (#243)
debezium/debezium.github.io: docs/connectors/sqlserver.asciidoc

:icons: font
:source-highlighter: highlight.js

> 

Reviewer: Can we leave in this note until 0.9.0.Final is out?

Author: @gunnarmorling  Well I exppected this to be merged with final version - hence I took a steap ahead

Reviewer: Ok, but why not merge this PR for the CR1 release? I.e. why hold it back?

== How the SQL Server connector works
TBD

[[schema-evolution]]
=== Database schema evolution

Debezium is able to capture schema changes over time.

> It is necessary that a database operator provides a co-operation when the schema is updated.

Reviewer: ```suggestion
Due to the way CDC is implemented in SQL Server, it is necessary to work in co-operation with a database operator in order to ensure the Debezium connector continues to produce data change events when the schema is updated.
```

> As was already mentioned before Debezium uses SQL Server's change data capture functionality.

Reviewer: ```suggestion
As was already mentioned before, Debezium uses SQL Server's change data capture functionality.
```

This means that SQL Server creates a capture table that contains all changes executed on the source table.
Unfortunately, the capture table is static and needs to be updated when the source table structure changes.

> This update is not done by the Debezium itself but must be executed by an operator with elevated privileges.

Reviewer: ```suggestion
This update is not done by the Debezium connector itself but must be executed by an operator with elevated privileges.
```

> There are generally two procedures how to execute the schema change

Reviewer: ```suggestion
There are generally two procedures how to execute the schema change:
```

> There are generally two procedures how to execute the schema change

Reviewer: ```suggestion
There are generally two procedures how to execute the schema change:
```

  * cold - this is executed when Debezium is stopped
  * hot - executed while Debezium is running

Both approaches have their own advantages and disadvantages.

[WARNING]
====

> In both cases, it is critically important to execute the procedure completely before a new schema update on the same source table is made.

Reviewer: ```suggestion
In both cases, it is critically important to execute the procedure completely before yet another schema update on the same source table is made.
```

It is thus recommended to execute all DDLs in a single batch so the procedure is done only once.
====

[NOTE]
====
Not all schema changes are supported when CDC is enabled for a source table.
One such exception identified is renaming the column - SQL Server will not allow executing the operation.
====

==== Cold schema update

This is the safest procedure but might not be feasible for applications with high-availability requirements.

> The operator should follow this sequence of steps

Reviewer: ```suggestion
The operator should follow this sequence of steps:
```

>   1. Suspend the application that generates the database records

Reviewer: ```suggestion
  1. Suspend the application(s) that write to the concerned database schema
```

>   2. Wait for Debezium to stream all unstreamed changes

Reviewer: ```suggestion
  2. Wait for Debezium to stream all changes
```

>   3. Stop Debezium connector

Reviewer: ```suggestion
  3. Stop the Debezium connector
```

  4. Apply all changes to the source table schema

>   5. Create a new capture table for the update source table using `sys.sp_cdc_enable_table` procedure with a unique value for parameter `@capture_instance`

Reviewer: ```suggestion
  5. Create a new capture table for the update source table using the `sys.sp_cdc_enable_table` procedure with a unique value for parameter `@capture_instance`
```

  6. Resume the application

>   7. Start Debezium connector

Reviewer: ```suggestion
  7. Start the Debezium connector
```

>   8. When Debezium starts streaming from the new capture table it is possible to drop the old one using `sys.sp_cdc_disable_table` stored procedure with parameter `@capture_instance` set to the old capture instance name

Reviewer: ```suggestion
  8. Once Debezium has started streaming from the new capture table, it is possible to drop the old one using the `sys.sp_cdc_disable_table` stored procedure with parameter `@capture_instance` set to the old capture instance name
```

Reviewer: How do I as a user know when the connector has begun to read from the new capture instance?

==== Hot schema update

The hot schema update does not require any downtime in application and data processing.

> The procedure itself is also much simpler than in case of cold schema update

Reviewer: ```suggestion
The procedure itself is also much simpler than in case of cold schema update:
```

  1. Apply all changes to the source table schema
  2. Create a new capture table for the update source table using `sys.sp_cdc_enable_table` procedure with a unique value for parameter `@capture_instance`

>   3. When Debezium starts streaming from the new capture table it is possible to drop the old one using `sys.sp_cdc_disable_table` stored procedure with parameter `@capture_instance` set to the old capture instance name

Reviewer: ```suggestion
  3. Once Debezium has started streaming from the new capture table, it is possible to drop the old one using the `sys.sp_cdc_disable_table` stored procedure with parameter `@capture_instance` set to the old capture instance name
```

The hot schema update has one drawback.
There is a period of time between the database schema update and creating the new capture instance.

> All changes that will arrive during this period will be captured by the old instance with the old structure.

Reviewer: ```suggestion
All changes that will arrive during this period will be captured by the old instance with the old structure.
For instance this means that in case of a newly added column any change event produced during this time will not yet contain a field for that new column.
```

If your application does not tolerate such a transition period we recommend to follow the cold schema update.

==== Example

> Deploy SQL Server based https://github.com/debezium/debezium-examples/tree/master/tutorial#using-sql-server[Debezium tutorial] to demonstrate the hot schema update.

Reviewer: ```suggestion
Let's deploy the SQL Server based https://github.com/debezium/debezium-examples/tree/master/tutorial#using-sql-server[Debezium tutorial] to demonstrate the hot schema update.
```

A column `phone_number` will be added to the `customers` table.

[source,shell]
----
# Start the database shell
docker-compose -f docker-compose-sqlserver.yaml exec sqlserver bash -c '/opt/mssql-tools/bin/sqlcmd -U sa -P $SA_PASSWORD -d testDB'
----

[source,sql]
----
-- Modify the source table schema
ALTER TABLE customers ADD phone_number VARCHAR(32);

-- Create the new capture instance
EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'customers', @role_name = NULL, @supports_net_changes = 0, @capture_instance = 'dbo_customers_v2';
GO

-- Insert new data
INSERT INTO customers(first_name,last_name,email,phone_number) VALUES ('John','Doe','john.doe@example.com', '+1-555-123456');
GO
----

> Kafka Connect log will contain messages

Reviewer: ```suggestion
Kafka Connect log will contain messages like these:
```

```

> connect_1    | 2019-01-17 10:11:14,924 INFO   ||  Multiple capture instances ChangeTable [captureInstance=dbo_customers, sourceTableId=testDB.dbo.customers, changeTableId=testDB.cdc.dbo_customers_CT, startLsn=00000024:00000d98:0036, changeTableObjectId=1525580473, stopLsn=00000025:00000ef8:0048] and ChangeTable [captureInstance=dbo_customers_v2, sourceTableId=testDB.dbo.customers, changeTableId=testDB.cdc.dbo_customers_v2_CT, startLsn=00000025:00000ef8:0048, changeTableObjectId=1749581271, stopLsn=NULL] present for the same table   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]

Reviewer: What is "ChangeTable"? A placeholder for something?

Author: @gunnarmorling It is the name of the object - see `toString()` method

Reviewer: Can you change this (in the code), so it reads like this:

> Multiple capture instances present for the same table:
> Capture instance "dbo_customers" (sourceTableId=..., ..., ...)
> Capture instance "dbo_customers_v2" (sourceTableId=...)

connect_1    | 2019-01-17 10:11:14,924 INFO   ||  Schema will be changed for ChangeTable [captureInstance=dbo_customers_v2, sourceTableId=testDB.dbo.customers, changeTableId=testDB.cdc.dbo_customers_v2_CT, startLsn=00000025:00000ef8:0048, changeTableObjectId=1749581271, stopLsn=NULL]   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]
...
connect_1    | 2019-01-17 10:11:33,719 INFO   ||  Migrating schema to ChangeTable [captureInstance=dbo_customers_v2, sourceTableId=testDB.dbo.customers, changeTableId=testDB.cdc.dbo_customers_v2_CT, startLsn=00000025:00000ef8:0048, changeTableObjectId=1749581271, stopLsn=NULL]   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]
```

> And there will be a new field in the schema and value of the message stored in Kafka

Reviewer: ```suggestion
Eventually, there will be a new field in the schema and value of the messages written to the Kafka topic.
```
